File: FCL/src/main/java/com/tungsten/fcl/ui/manage/WorldListPage.java
package com.tungsten.fcl.ui.manage;

import android.app.Activity;
import android.content.Context;
import android.net.Uri;
import android.view.View;
import android.widget.ListView;

import com.tungsten.fcl.R;
import com.tungsten.fcl.setting.Profile;
import com.tungsten.fcl.util.AndroidUtils;
import com.tungsten.fcl.util.RequestCodes;
import com.tungsten.fclauncher.utils.FCLPath;
import com.tungsten.fclcore.fakefx.beans.property.BooleanProperty;
import com.tungsten.fclcore.fakefx.beans.property.ListProperty;
import com.tungsten.fclcore.fakefx.beans.property.SimpleBooleanProperty;
import com.tungsten.fclcore.fakefx.beans.property.SimpleListProperty;
import com.tungsten.fclcore.fakefx.collections.FXCollections;
import com.tungsten.fclcore.game.World;
import com.tungsten.fclcore.task.Schedulers;
import com.tungsten.fclcore.task.Task;
import com.tungsten.fclcore.util.Logging;
import com.tungsten.fcllibrary.browser.FileBrowser;
import com.tungsten.fcllibrary.browser.options.LibMode;
import com.tungsten.fcllibrary.browser.options.SelectionMode;
import com.tungsten.fcllibrary.component.dialog.FCLAlertDialog;
import com.tungsten.fcllibrary.component.ui.FCLCommonPage;
import com.tungsten.fcllibrary.component.view.FCLButton;
import com.tungsten.fcllibrary.component.view.FCLCheckBox;
import com.tungsten.fcllibrary.component.view.FCLProgressBar;
import com.tungsten.fcllibrary.component.view.FCLUILayout;

import java.io.File;
import java.io.IOException;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.InvalidPathException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.logging.Level;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class WorldListPage extends FCLCommonPage implements ManageUI.VersionLoadable, View.OnClickListener {

    private final ListProperty<WorldListItem> itemsProperty = new SimpleListProperty<>(FXCollections.observableArrayList());

    private final BooleanProperty showAll = new SimpleBooleanProperty(this, "showAll", false);

    private Path savesDir;
    private List<World> worlds;
    private Profile profile;
    private String id;
    private String gameVersion;

    private FCLCheckBox showAllCheckBox;
    private FCLButton addButton;
    private FCLButton refreshButton;
    private ListView listView;
    private FCLProgressBar progressBar;

    public WorldListPage(Context context, int id, FCLUILayout parent, int resId) {
        super(context, id, parent, resId);

        create();
    }

    public void create() {
        showAllCheckBox = findViewById(R.id.show_all);
        addButton = findViewById(R.id.add);
        refreshButton = findViewById(R.id.refresh);
        listView = findViewById(R.id.list);
        progressBar = findViewById(R.id.progress);

        showAll.addListener(e -> {
            if (worlds != null)
                itemsProperty.setAll(worlds.stream()
                        .filter(world -> isShowAll() || world.getGameVersion() == null || world.getGameVersion().equals(gameVersion))
                        .map(it -> new WorldListItem(getContext(), getActivity(), getParent(), it)).collect(Collectors.toList()));
        });

        showAllCheckBox.addCheckedChangeListener();
        showAllCheckBox.checkProperty().bindBidirectional(showAllProperty());
        addButton.setOnClickListener(this);
        refreshButton.setOnClickListener(this);

        WorldListAdapter adapter = new WorldListAdapter(getContext());
        adapter.listProperty().bind(itemsProperty);
        listView.setAdapter(adapter);
    }

    @Override
    public Task<?> refresh(Object... param) {
        return null;
    }

    @Override
    public void loadVersion(Profile profile, String version) {
        this.profile = profile;
        this.id = version;
        this.savesDir = profile.getRepository().getRunDirectory(id).toPath().resolve("saves");
        refresh();
    }

    @Override
    public void onClick(View v) {
        if (v == addButton) {
            add();
        }
        if (v == refreshButton) {
            refresh();
        }
    }

    public CompletableFuture<?> refresh() {
        if (profile == null || id == null)
            return CompletableFuture.completedFuture(null);

        setLoading(true);
        return CompletableFuture
                .runAsync(() -> gameVersion = profile.getRepository().getGameVersion(id).orElse(null))
                .thenApplyAsync(unused -> {
                    try (Stream<World> stream = World.getWorlds(savesDir)) {
                        return stream.parallel().collect(Collectors.toList());
                    }
                })
                .whenCompleteAsync((result, exception) -> {
                    worlds = result;
                    setLoading(false);
                    if (exception == null)
                        itemsProperty.setAll(result.stream()
                                .filter(world -> isShowAll() || world.getGameVersion() == null || world.getGameVersion().equals(gameVersion))
                                .map(it -> new WorldListItem(getContext(), getActivity(), getParent(), it)).collect(Collectors.toList()));
                }, Schedulers.androidUIThread());
    }

    private void setLoading(boolean loading) {
        progressBar.setVisibility(loading ? View.VISIBLE : View.GONE);
        listView.setVisibility(loading ? View.GONE : View.VISIBLE);
        showAllCheckBox.setEnabled(!loading);
        addButton.setEnabled(!loading);
        refreshButton.setEnabled(!loading);
    }

    public void add() {
        FileBrowser.Builder builder = new FileBrowser.Builder(getContext());
        builder.setLibMode(LibMode.FILE_CHOOSER);
        builder.setSelectionMode(SelectionMode.SINGLE_SELECTION);
        ArrayList<String> suffix = new ArrayList<>();
        suffix.add(".zip");
        builder.setSuffix(suffix);
        builder.create().browse(getActivity(), RequestCodes.SELECT_WORLD_CODE, ((requestCode, resultCode, data) -> {
            if (requestCode == RequestCodes.SELECT_WORLD_CODE && resultCode == Activity.RESULT_OK && data != null) {
                String path = FileBrowser.getSelectedFiles(data).get(0);
                Uri uri = Uri.parse(path);
                if (AndroidUtils.isDocUri(uri)) {
                    path = AndroidUtils.copyFileToDir(getActivity(), uri, new File(FCLPath.CACHE_DIR));
                }
                File file = new File(path);
                installWorld(file);
            }
        }));
    }

    private void installWorld(File zipFile) {
        // Only accept one world file because user is required to confirm the new world name
        // Or too many input dialogs are popped.
        FCLAlertDialog.Builder builder = new FCLAlertDialog.Builder(getContext());
        builder.setCancelable(false);
        builder.setAlertLevel(FCLAlertDialog.AlertLevel.INFO);
        builder.setMessage(getContext().getString(R.string.world_add));
        FCLAlertDialog installDialog = builder.create();
        installDialog.show();
        Task.supplyAsync(() -> new World(zipFile.toPath()))
                .whenComplete(Schedulers.androidUIThread(), world -> {
                    installDialog.dismiss();
                    WorldNameDialog dialog = new WorldNameDialog(getContext(), world.getWorldName(), (name, resolve, reject) -> Task.runAsync(() -> world.install(savesDir, name))
                            .whenComplete(Schedulers.androidUIThread(), () -> {
                                itemsProperty.add(new WorldListItem(getContext(), getActivity(), getParent(), new World(savesDir.resolve(name))));
                                resolve.run();
                            }, e -> {
                                if (e instanceof FileAlreadyExistsException)
                                    reject.accept(AndroidUtils.getLocalizedText(getContext(), "world_import_failed", getContext().getString(R.string.world_import_already_exists)));
                                else if (e instanceof IOException && e.getCause() instanceof InvalidPathException)
                                    reject.accept(AndroidUtils.getLocalizedText(getContext(), getContext().getString(R.string.install_new_game_malformed)));
                                else
                                    reject.accept(AndroidUtils.getLocalizedText(getContext(), e.getClass().getName() + ": " + e.getLocalizedMessage()));
                            }).start());
                    dialog.show();
                }, e -> {
                    installDialog.dismiss();
                    Logging.LOG.log(Level.WARNING, "Unable to parse world file " + zipFile, e);
                    FCLAlertDialog.Builder builder1 = new FCLAlertDialog.Builder(getContext());
                    builder1.setCancelable(false);
                    builder1.setAlertLevel(FCLAlertDialog.AlertLevel.ALERT);
                    builder1.setMessage(getContext().getString(R.string.world_import_invalid));
                    builder1.setNegativeButton(getContext().getString(com.tungsten.fcllibrary.R.string.dialog_positive), null);
                    builder1.create().show();
                }).start();
    }

    public boolean isShowAll() {
        return showAll.get();
    }

    public BooleanProperty showAllProperty() {
        return showAll;
    }

    public void setShowAll(boolean showAll) {
        this.showAll.set(showAll);
    }
}

File: FCL/src/main/java/com/tungsten/fcl/ui/manage/WorldListPage.kt
package com.tungsten.fcl.ui.manage

import android.content.Context
import android.view.View
import android.widget.Toast
import androidx.core.net.toUri
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.mio.util.showErrorDialog
import com.tungsten.fcl.R
import com.tungsten.fcl.activity.MainActivity
import com.tungsten.fcl.databinding.PageManageWorldBinding
import com.tungsten.fcl.setting.Profile
import com.tungsten.fcl.ui.manage.ManageUI.VersionLoadable
import com.tungsten.fcl.util.AndroidUtils
import com.tungsten.fclauncher.utils.FCLPath
import com.tungsten.fclcore.fakefx.beans.Observable
import com.tungsten.fclcore.fakefx.beans.property.BooleanProperty
import com.tungsten.fclcore.fakefx.beans.property.ListProperty
import com.tungsten.fclcore.fakefx.beans.property.SimpleBooleanProperty
import com.tungsten.fclcore.fakefx.beans.property.SimpleListProperty
import com.tungsten.fclcore.fakefx.collections.FXCollections
import com.tungsten.fclcore.game.World
import com.tungsten.fclcore.task.Task
import com.tungsten.fclcore.util.Logging
import com.tungsten.fcllibrary.component.dialog.EditDialog
import com.tungsten.fcllibrary.component.dialog.FCLAlertDialog
import com.tungsten.fcllibrary.component.ui.FCLPage
import com.tungsten.fcllibrary.component.view.FCLUILayout
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import java.io.File
import java.io.IOException
import java.nio.file.FileAlreadyExistsException
import java.nio.file.Files
import java.nio.file.InvalidPathException
import java.nio.file.Path
import java.util.logging.Level
import java.util.stream.Collectors
import kotlin.coroutines.resume
import kotlin.io.path.pathString

class WorldListPage(context: Context?, id: Int, resId: Int) : FCLPage(context, id, resId), VersionLoadable, View.OnClickListener {
    private val itemsProperty: ListProperty<WorldListItem> =
        SimpleListProperty(FXCollections.observableArrayList())

    private val showAll: BooleanProperty = SimpleBooleanProperty(this, "showAll", false)

    private lateinit var savesDir: Path
    private var worlds: MutableList<World?> = mutableListOf()
    private var profile: Profile? = null
    private var id: String? = null
    private var gameVersion: String? = null
    private lateinit var binding: PageManageWorldBinding

    init {
        create()
    }

    fun create() {
        binding = PageManageWorldBinding.bind(contentView)

        showAll.addListener { _: Observable? ->
            itemsProperty.setAll(
                worlds.stream()
                    .filter { world: World? -> isShowAll() || world!!.gameVersion == null || world.gameVersion == gameVersion }
                    .map { it: World? ->
                        WorldListItem(
                            context,
                            activity,
                            it
                        )
                    }.collect(
                        Collectors.toList()
                    )
            )
        }

        binding.showAll.addCheckedChangeListener()
        binding.showAll.checkProperty().bindBidirectional(showAll)
        binding.add.setOnClickListener(this)
        binding.refresh.setOnClickListener(this)
        binding.fixPrivate.setOnClickListener(this)

        val adapter = WorldListAdapter(context)
        adapter.listProperty().bind(itemsProperty)
        binding.recyclerView.setLayoutManager(LinearLayoutManager(context))
        binding.recyclerView.setAdapter(adapter)
    }

    override fun refresh(vararg param: Any?): Task<*>? {
        return null
    }

    override fun loadVersion(profile: Profile, version: String?) {
        this.profile = profile
        this.id = version
        this.savesDir = profile.repository.getRunDirectory(id).toPath().resolve("saves")
        refresh()
    }

    override fun onClick(v: View?) {
        when (v) {
            binding.add -> add()
            binding.refresh -> refresh()
            binding.fixPrivate -> {
                // use 关闭 Files.walk 的目录流，避免文件描述符泄漏
                Files.walk(savesDir).use { stream ->
                    stream.forEach { path ->
                        Files.setAttribute(
                            path,
                            "unix:mode",
                            1535
                        )
                    }
                }
                Toast.makeText(context, R.string.message_success, Toast.LENGTH_LONG).show()
            }
        }
    }

    fun refresh() {
        if (profile == null || id == null) return
        setLoading(true)
        MainActivity.getInstance().lifecycleScope.launch {
            val result = runCatching {
                withContext(Dispatchers.IO) {
                    gameVersion = profile!!.repository.getGameVersion(id).orElse(null)
                    World.getWorlds(
                        savesDir
                    )
                }
            }.getOrElse {
                worlds.clear()
                return@launch
            }
            setLoading(false)
            worlds.addAll(result)
            itemsProperty.setAll(
                result.stream()
                    .filter { isShowAll() || it.gameVersion == null || it.gameVersion == gameVersion }
                    .map {
                        WorldListItem(
                            context,
                            activity,
                            it
                        )
                    }.collect(
                        Collectors.toList()
                    )
            )
            if (savesDir.pathString.startsWith(FCLPath.PRIVATE_COMMON_DIR) or savesDir.pathString.contains(
                    "/Android/data/${context.applicationInfo.packageName}/"
                )
            ) {
                if (worlds.isNotEmpty())
                    binding.fixPrivate.visibility = View.VISIBLE
            } else {
                binding.fixPrivate.visibility = View.GONE
            }
        }
    }

    private fun setLoading(loading: Boolean) {
        binding.progress.visibility = if (loading) View.VISIBLE else View.GONE
        binding.recyclerView.visibility = if (loading) View.GONE else View.VISIBLE
        binding.showAll.setEnabled(!loading)
        binding.add.setEnabled(!loading)
        binding.refresh.setEnabled(!loading)
    }

    fun add() {
        MainActivity.getInstance().fileLauncher.launchSingleSelection(null, listOf(".zip")) {
            var path = it?.get(0) ?: return@launchSingleSelection
            val uri = path.toUri()
            if (AndroidUtils.isDocUri(uri)) {
                path =
                    AndroidUtils.copyFileToDir(activity, uri, File(FCLPath.CACHE_DIR))
            }
            val file = File(path)
            installWorld(file)
        }
    }

    private fun installWorld(zipFile: File) {
        // Only accept one world file because user is required to confirm the new world name
        // Or too many input dialogs are popped.
        val builder = FCLAlertDialog.Builder(context)
        builder.setCancelable(false)
        builder.setAlertLevel(FCLAlertDialog.AlertLevel.INFO)
        builder.setMessage(context.getString(R.string.world_add))
        val installDialog = builder.create()
        installDialog.show()
        MainActivity.getInstance().lifecycleScope.launch(Dispatchers.Main) {
            val world = runCatching {
                withContext(Dispatchers.IO) {
                    World(zipFile.toPath())
                }
            }.getOrElse {
                installDialog.dismiss()
                Logging.LOG.log(Level.WARNING, "Unable to parse world file $zipFile", it)
                val builder1 = FCLAlertDialog.Builder(context)
                builder1.setCancelable(false)
                builder1.setAlertLevel(FCLAlertDialog.AlertLevel.ALERT)
                builder1.setMessage(context.getString(R.string.world_import_invalid))
                builder1.setNegativeButton(
                    context.getString(com.tungsten.fcl.R.string.dialog_positive),
                    null
                )
                builder1.create().show()
                return@launch
            }
            installDialog.dismiss()
            val name = showEditDialog(world.worldName) ?: return@launch
            runCatching {
                withContext(Dispatchers.IO) {
                    world.install(savesDir, name)
                }
            }.onFailure {
                val error = when (it) {
                    is FileAlreadyExistsException -> AndroidUtils.getLocalizedText(
                        context,
                        "world_import_failed",
                        context.getString(R.string.world_import_already_exists)
                    )

                    is IOException if it.cause is InvalidPathException -> AndroidUtils.getLocalizedText(
                        context,
                        context.getString(R.string.install_new_game_malformed)
                    )

                    else -> AndroidUtils.getLocalizedText(
                        context,
                        it.javaClass.getName() + ": " + it.localizedMessage
                    )
                }
                showErrorDialog(context, error)
            }.onSuccess {
                itemsProperty.add(
                    WorldListItem(
                        context,
                        activity,
                        World(savesDir.resolve(name))
                    )
                )
            }
        }
    }

    private suspend fun showEditDialog(worldName: String): String? =
        suspendCancellableCoroutine {
            val dialog = EditDialog(context, worldName) { name ->
                it.resume(name)
            }
            dialog.onCancelListener = {
                it.resume(null)
            }
            dialog.show()
            it.invokeOnCancellation {
                dialog.dismiss()
            }
        }

    fun isShowAll(): Boolean {
        return showAll.get()
    }
}
